src/cloud/digitalocean/provisioner.ts

```diff
@@ -182,38 +191,52 @@ class Provisioner {
   }
 
   /**
-   * Try to retrieve SSH keys from storage.
-   * If not found, generate new ones and store
-   * @param {String} name name of the key (usually same as name of VM later)
-   * @return {Promise.<KeyPair>} the SSH keys retrieved
-   * {
-   *    public: "...",
-   *    private: "..."
-   * }
+   * Retrieves an SSH keypair from storage, generating and saving
+   * a new keypair if none is found.
+   * @param {String} name name of the key
+   * @return {Promise.<KeyPair>}
+   * @throws if there is any problem loading from or saving to
+   * storage or if a keypair cannot be generated
    */
   private getSshKey_ = (name: string) : Promise<KeyPair> => {
-    var storage = freedom["core.storage"]();
-    return new Promise((F, R) => {
-      var result: KeyPair = <KeyPair>{};
-      Promise.all([
-        storage.get("DigitalOcean-" + name + "-PublicKey"),
-        storage.get("DigitalOcean-" + name + "-PrivateKey")
-      ]).then((val: string[]) => {
-        if (val[0] === null ||
-            val[1] === null) {
-          result = Provisioner.generateKeyPair_();
-          storage.set("DigitalOcean-" + name + "-PublicKey", result.public);
-          storage.set("DigitalOcean-" + name + "-PrivateKey", result.private);
-          this.sendStatus_("SSHKEY_GENERATED");
-        } else {
-          result.public = val[0];
-          result.private = val[1];
-          this.sendStatus_("SSHKEY_RETRIEVED");
-        }
-        F(result);
-      }).catch((err: Error) => {
-        console.error("storage error: " + JSON.stringify(err));
-        R(err);
+    const publicKeyIndex = 'DigitalOcean-' + name + '-PublicKey';
+    const privateKeyIndex = 'DigitalOcean-' + name + '-PrivateKey';
+    const storage = freedom['core.storage']();
+    return Promise.all([
+        storage.get(publicKeyIndex),
+        storage.get(privateKeyIndex)
+    ]).then((results: string[]) => {
+      if (results[0] !== null && results[1] !== null) {
+        this.sendStatus_('SSHKEY_RETRIEVED');
+        return {
+          public: results[0],
+          private: results[1]
+        };
+      }
+
+      this.sendStatus_('GENERATING_SSHKEY');
+      try {
+        const result = Provisioner.generateKeyPair_();
+        this.sendStatus_('SSHKEY_GENERATED');
+        return Promise.all([
+          storage.set(publicKeyIndex, result.public),
+          storage.set(privateKeyIndex, result.private)
+        ]).then((ignored: any) => {
```

Comment:
I can, and it will compile, but I'd rather make it explicit that while we know there's a returned object, we don't need it. Plus, TypeScript has been getting better and better at recognising things like extra fields in objects -- this may start raising an error in future versions.